crates/ink/src/contract_ref.rs

```diff
@@ -14,6 +14,115 @@
+///         #[ink(message)]
+///         pub fn change_account_id_1(&mut self, new_erc20: AccountId) {
+///             self.erc20 = new_erc20.into();
+///         }
+///
+///         /// The example of converting `AccountId` into `contract_ref!` explicitly.
+///         #[ink(message)]
+///         pub fn change_account_id_2(&mut self, new_erc20: AccountId) {
+///             let erc20: contract_ref!(Erc20) = new_erc20.into();
+///             self.erc20 = erc20;
+///         }
+///
+///         /// The example of converting `AccountId` into `contract_ref!` explicitly
+///         /// with custom environment.
+///         #[ink(message)]
+///         pub fn change_account_id_3(&mut self, new_erc20: AccountId) {
+///             let erc20: contract_ref!(Erc20, ink::env::DefaultEnvironment) = new_erc20.into();
+///             self.erc20 = erc20;
+///         }
+///
+///         /// The example of converting `AccountId` into `contract_ref!` explicitly
+///         /// with custom environment from the associated type.
+///         #[ink(message)]
+///         pub fn change_account_id_4(&mut self, new_erc20: AccountId) {
+///             let erc20: contract_ref!(Erc20, <Caller as ink::env::ContractEnv>::Env) = new_erc20.into();
+///             self.erc20 = erc20;
+///         }
+///
+///         /// The example of how to do common calls.
+///         #[ink(message)]
+///         pub fn total_supply_1(&self) -> Balance {
+///             self.erc20.total_supply()
+///         }
+///
+///         /// The example of how to use the call builder with `contract_ref!`.
+///         #[ink(message)]
+///         pub fn total_supply_2(&self) -> Balance {
+///             use ink::codegen::TraitCallBuilder;
+///             let erc20_builder = self.erc20.call();
+///             let err: ink::env::Result<ink::primitives::MessageResult<Balance>> =
+///                 erc20_builder.total_supply().transferred_value(0).try_invoke();
+///             err
+///                 .expect("The cross-contract call should be executed without ink::env::Error")
+///                 .expect("The cross-contract call should be executed without ink::primitives::LangError")
+///         }
+///
+///         /// The example of how to do common calls and convert the `contract_ref!` into `AccountId`.
+///         #[ink(message)]
+///         pub fn transfer_to_erc20(&mut self, amount: Balance) -> bool {
+///             let erc20_as_account_id = self.erc20.as_ref().clone();
+///             self.erc20.transfer(amount, erc20_as_account_id)
+///         }
+///     }
+/// }
+/// ```
+#[macro_export]
+macro_rules! contract_ref {
+    // The case of the default `Environment`
+    ( $trait_path:path ) => {
+        $crate::contract_ref!($trait_path, $crate::env::DefaultEnvironment)
+    };
+    // The case of the custom `Environment`
+    ( $trait_path:path, $env:ty ) => {
+        <<$crate::reflect::TraitDefinitionRegistry<$env> as $trait_path>
```

Comment:
```suggestion
        <<$crate::reflect::TraitDefinitionRegistry<Environment> as $trait_path>
```
This would use the same environment as the containing contract. Assumes this is always used in the context of `#[ink::contract]` of course.

Reply:
You can use `contract_ref!` anywhere. It is not bounded to the `#[ink::contract]` and it is the primary feature. You can use it in any file anywhere=) It returns a plain rust type.

The `contract_ref!` works with `ink::env::DefaultEnvironment` by default, so the user should write `contract_ref!(Erc20, Environment)` to use the environment from the contract.

I did that because `DefaultEnvironment` is always available, and the macro will not throw an error outside of the `#[ink::contract]`.

But we can change the default behavior=) The `contract_ref!` could use `Environment` instead of the `DefaultEnvironment`. Then outside of the `#[ink::contract]` the user should use `contract_ref!(Erc20, DefaultEnvironment)`(some environment).

Reply:
> But we can change the default behavior=) The contract_ref! could use Environment instead of the DefaultEnvironment. Then outside of the #[ink::contract] the user should use contract_ref!(Erc20, DefaultEnvironment)(some environment).

It would be nice, but maybe it's too early in the morning but I can't think of how we would achieve this?

Reply:
Very easy=) 
<img width="1505" alt="image" src="https://user-images.githubusercontent.com/18346821/221143654-b05bf6c7-f439-4748-895c-565aeca36466.png">

Reply:
Oh I see so you just mean to use `Environment` by default, and if it is outside the contract then the user will just get the compilation error that it can't find this type. I think that is reasonable, as long as it is well documented. It would save users who use their own custom environment for having to specify it everywhere when used inside a contract.

Reply:
> Very easy=)

Yes I was stupidly thinking you would detect the context from within the macro. But yes that makes sense.

If it is well documented I would go this way, assuming that most people will use it inside `#[ink::contract]`